File: python/713.subarray-product-less-than-k.py
from typing import List


class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        res = 0
        prod = 1
        p, q = 0, 0

        while q < len(nums):

            prod *= nums[q]

            while p <= q and prod >= k:
                prod //= nums[p]
                p += 1

            res += q - p + 1
            q += 1

        return res

File: cpp/713.subarray-product-less-than-k.cpp
class Solution {
public:
  int numSubarrayProductLessThanK(vector<int> &nums, int k) {
    int prod = 1;
    int p = 0;
    int res = 0;

    for (int q = 0; q < nums.size(); q++) {
      prod *= nums[q];

      while (p <= q && prod >= k) {
        prod /= nums[p];
        p++;
      }

      res += (q - p + 1);
    }

    return res;
  }
};